wallet/src/main/scala/org/bitcoins/wallet/models/SpendingInfoDAO.scala

```diff
@@ -47,6 +45,143 @@ case class SpendingInfoDAO()(implicit
+    ScriptPubKeyDAO#ScriptPubKeyTable] = {
+    ScriptPubKeyDAO().table
+  }
+
+  def create(si: SpendingInfoDb): Future[SpendingInfoDb] = {
+    val query =
+      table.returning(table.map(_.id)).into((t, id) => t.copyWithId(id = id))
+
+    val actions = for {
+      spkOpt <-
+        spkTable
+          .filter(_.scriptPubKey === si.output.scriptPubKey)
+          .result
+          .headOption
+      utxo: UTXORecord <- spkOpt match {
+        case Some(foundSpk) =>
+          val utxo = UTXORecord.fromSpendingInfoDb(si, foundSpk.id.get)
+          query += utxo
+        case None =>
+          (for {
+            newSpkId <-
+              (spkTable returning spkTable.map(_.id)) += (ScriptPubKeyDb(
+                si.output.scriptPubKey))
+          } yield {
+            val utxo = UTXORecord.fromSpendingInfoDb(si, newSpkId)
+            query += utxo
+          }).flatten
+      }
+      spk <-
+        spkTable
+          .filter(_.id === utxo.scriptPubKeyId)
+          .result
+          .headOption
+    } yield (utxo, spk)
+
+    safeDatabase
+      .run(actions.transactionally)
+      .map {
+        case (utxo, Some(spk)) => utxo.toSpendingInfoDb(spk.scriptPubKey)
+        case _                 => throw new SQLException("unexpected result")
+      }
+  }
+
+  def upsertAllSpendingInfoDb(
+      ts: Vector[SpendingInfoDb]): Future[Vector[SpendingInfoDb]] = {
+    FutureUtil.foldLeftAsync(Vector.empty[SpendingInfoDb], ts)((acc, si) =>
+      upsert(si).map(res => acc :+ res))
+  }
+
+  def updateAllSpendingInfoDb(
+      ts: Vector[SpendingInfoDb]): Future[Vector[SpendingInfoDb]] = {
+    FutureUtil.foldLeftAsync(Vector.empty[SpendingInfoDb], ts)((acc, si) =>
+      update(si).map(res => acc :+ res))
+  }
+
+  def update(si: SpendingInfoDb): Future[SpendingInfoDb] = {
+    val actions = for {
+      spkOpt <-
+        spkTable
+          .filter(_.scriptPubKey === si.output.scriptPubKey)
+          .result
+          .headOption
+      _ <- spkOpt match {
+        case Some(foundSpk) =>
+          val utxo = UTXORecord.fromSpendingInfoDb(si, foundSpk.id.get)
+          table.filter(_.id === utxo.id).update(utxo)
+        case None =>
+          (for {
+            newSpkId <-
+              (spkTable returning spkTable.map(_.id)) += (ScriptPubKeyDb(
+                si.output.scriptPubKey))
+          } yield {
+            val utxo = UTXORecord.fromSpendingInfoDb(si, newSpkId)
+            table.filter(_.id === utxo.id).update(utxo)
+          }).flatten
+      }
+      utxo <- table.filter(_.id === si.id.get).result.headOption
+      spk <-
+        spkTable
+          .filter(_.scriptPubKey === si.output.scriptPubKey)
+          .result
+          .headOption
+    } yield (utxo, spk)
+    safeDatabase
+      .run(actions.transactionally)
+      .map {
+        case (Some(utxo), Some(spk)) => utxo.toSpendingInfoDb(spk.scriptPubKey)
+        case _                       => throw new SQLException("unexpected result")
```

Comment:
I'd sum up all those messages as `Shut up compiler!`. If any of these three cases occur, that means the underlying DBMS has failed catastrophically, so that it cannot process DB transactions in a transactional manner and remains silent about those errors.